File: Iteration_2/src/main/java/DepartmentScheduler.java
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;

public class DepartmentScheduler extends Staff {
    private ArrayList<CourseSection> courseSections;
    private ArrayList<String> allClassrooms;
    private ArrayList<String> allTimeIntervals;

    public DepartmentScheduler() {
        super();
        this.allTimeIntervals = initTimeIntervals();
    }
    public DepartmentScheduler(String name, String surname, ArrayList<CourseSection> courseSections, ArrayList<String> allClassrooms) {
        super(name, surname);
        this.allTimeIntervals = initTimeIntervals();
        this.courseSections = courseSections;
        this.allClassrooms = allClassrooms;
    }

    // Constructor for Department Scheduler role.
    public DepartmentScheduler(String name, String surname, Date birthdate, char gender, String ssn,
            ArrayList<CourseSection> courseSections, ArrayList<String> allClassrooms) {
        super(name, surname, birthdate, gender, ssn);
        this.courseSections = courseSections;
        this.allClassrooms = allClassrooms;
        this.allTimeIntervals = initTimeIntervals();
    }

    public ArrayList<String> initTimeIntervals() {
        ArrayList<String> allTimeIntervals = new ArrayList<>();
        allTimeIntervals.add("8:30-9:20");
        allTimeIntervals.add("9:30-10:20");
        allTimeIntervals.add("10:30-11:20");
        allTimeIntervals.add("11:30-12:20");
        allTimeIntervals.add("12:30-13:20");
        allTimeIntervals.add("13:30-14:20");
        allTimeIntervals.add("14:30-15:20");
        allTimeIntervals.add("15:30-16:20");
        return allTimeIntervals;
    }



    // Assign time slot for specific section.
    public void assignTimeSlotToSection(CourseSection courseSection, TimeSlot timeSlot) {
        try {
            if (courseSection == null || timeSlot == null) {
                throw new IllegalArgumentException("CourseSection or TimeSlot cannot be null.");
            }
            timeSlot.assignTimeSlot(courseSection);
        } catch (Exception e) {
            System.err.println("Error assigning time slot to section: " + e.getMessage());
        }
    }

    // Assign the lecturer to specific section.
    public void assignLecturerToSection(CourseSection courseSection, Lecturer lecturer) {
        try {
            if (courseSection == null || lecturer == null) {
                throw new IllegalArgumentException("CourseSection or Lecturer cannot be null.");
            }
            if (handleLecturerConflict(lecturer, courseSection))
                courseSection.setLecturer(lecturer);
        } catch (Exception e) {
            System.err.println("Error assigning lecturer to section: " + e.getMessage());
        }
    }

    // Change capacity of specific section.
    public void manageCapacity(CourseSection courseSection, int newCapacity) {
        try {
            if (courseSection == null) {
                throw new IllegalArgumentException("CourseSection cannot be null.");
            }
            if (newCapacity < courseSection.getCapacity()) {
                throw new IllegalArgumentException("New capacity cannot be smaller than the old capacity.");
            }

            int oldCapacity = courseSection.getCapacity();
            courseSection.setCapacity(newCapacity);

            // Manage waitlist based on new capacity.
            manageWaitlist(courseSection, newCapacity - oldCapacity);
        } catch (Exception e) {
            System.err.println("Error managing capacity: " + e.getMessage());
        }
    }

    // Make changes in waitlist. Remove student from the waitlist of the Course
    // Section.
    public void manageWaitlist(CourseSection courseSection, int size) {
        try {
            if (courseSection == null) {
                throw new IllegalArgumentException("CourseSection cannot be null.");
            }
            ArrayList<Student> studentsToSection = courseSection.getWaitList();

            for (int i = 0; i < size && !studentsToSection.isEmpty(); i++) {
                Student student = studentsToSection.get(0);

                if (student.getTranscript() == null) {
                    throw new NullPointerException("Student's transcript cannot be null.");
                }

                // Add course to student's transcript.
                student.getTranscript().addCurrentCourse(courseSection.getParentCourse());
                int sizeNew = student.getTranscript().getWaitedCourses().size();
                for(int j = 0; j < sizeNew; j++){
                    if(student.getTranscript().getWaitedCourses().get(j).getCourseId().equals(courseSection.getParentCourse().getCourseId())){
                        student.getTranscript().deleteFromWaitedCourse(student.getTranscript().getWaitedCourses().get(j));
                    }
                }
                student.getTranscript().addCurrentSection(courseSection);
                student.getTranscript().deleteFromWaitedSections(courseSection);

                // Remove student from the waitlist.
                studentsToSection.remove(0);
            }

            courseSection.setWaitList(studentsToSection);
        } catch (Exception e) {
            System.err.println("Error managing waitlist: " + e.getMessage());
            e.printStackTrace();
        }
    }

    // TİMESLOTU NULL OLAN BİR SECTION SEÇERSE;
    // -ÖNCE SAAT SOR. SONRA CLASSROOM
    // UPDATE SECTION SEÇERSE; (O SECTION'IN ÖNCE SAAT KISMINI NULL YAP Kİ O DA
    // SEÇENEK LİSTESİNDE GÖZÜKSÜN)

    // DepartmentSchedulerInterface önce CourseSection listesinden seçim yapılacak.
    // Seçilen ders yeni oluşturulduysa("TimeSlot kısmı null ise veya capacity
    // girilmemişse diye kontrol edilebilir")
    // -handleTimeConflict methodunu çağır, (bu method seçilebilecek zaman
    // aralıklarını döndürüyor)
    // -handleClassroomConflict methodunu çağır, (bu method seçilen zamana göre
    // seçilebilecek sınıfları döndürüyor)
    // -Kullanıcının seçtiği time interval'lar ile classroomu kullanarak TimeSlot
    // objesi oluştur ve courseSection.setTimeSlot ile eşleştir.

    // Check time conflict for CourseSections whichs is in the same semester.
    // Semester Courses are needed.
    public ArrayList<String> handleTimeConflict(ArrayList<CourseSection> semesterCourses, String day) {
        // Saatleri seçmesi istenecek. Birden fazla saat seçilebilir. Bu method aynı
        // yıldaki diğer tüm sectionların saatine bakmalı ve saat çakışmalarını
        // engellemeli
        ArrayList<String> availableTimeIntervals = new ArrayList<>(allTimeIntervals);

        try {
            // semesterCourses null ise hata fırlat
            if (semesterCourses == null) {
                throw new IllegalArgumentException("Semester courses list cannot be null.");
            }
    
            // SemesterCourses içindeki TimeSlots'ları kontrol et
            for (int i = 0; i < semesterCourses.size(); i++) {
                ArrayList<TimeSlot> timeSlots = semesterCourses.get(i).getTimeSlots();
    
                // Gün ve zaman aralığı kontrolü
                for (int k = 0; k < timeSlots.size(); k++) {
                    String timeInterval = timeSlots.get(k).getTimeInterval();
                    String sectionDay = timeSlots.get(k).getDay();
                    if (sectionDay.equals(day)) {
                        availableTimeIntervals.remove(timeInterval);
                    }
                }
            }
        } catch (IllegalArgumentException e) {
            System.out.println("Error: " + e.getMessage());
            System.exit(1); // Hata durumunda programı sonlandır
        } catch (Exception e) {
            System.out.println("An unexpected error occurred: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    
        return availableTimeIntervals;
    }
    

    // Returns x'th semester courses as ArrayList.
    public ArrayList<CourseSection> semesterXCourses(int x) {
        ArrayList<CourseSection> semesterXCourses = new ArrayList<>();

        for (int i = 0; i < courseSections.size(); i++) {
            if (courseSections.get(i).getParentCourse().getSemester() == x) {
                semesterXCourses.add(courseSections.get(i));
            }
        }

        return semesterXCourses;
    }

    // Check classroom conflict for CourseSection.
    public ArrayList<String> handleClassroomConflict(String day, String timeInterval) {
       // Return classrooms which can be selected. So, decide whether the classroom is selected before or not.
        ArrayList<String> availableClassrooms = new ArrayList<>(allClassrooms);

        for (int i = 0; i < courseSections.size(); i++) {
            ArrayList<TimeSlot> timeSlots = courseSections.get(i).getTimeSlots();
            for (int k = 0; k < timeSlots.size(); k++) {
                TimeSlot slot = timeSlots.get(k);
                // Gün ve zaman aralığını kontrol et
                if (slot.getDay().equals(day) && slot.getTimeInterval().equals(timeInterval)) {
                    String whichClassroom = findWhichClassroom(slot.getClassroom());
                    availableClassrooms.remove(whichClassroom);
                }
            }
        }

        return availableClassrooms;
    }

    public String findWhichClassroom(String classroom) {
        String classroomFound = null;
        try {
            for (int i = 0; i < allClassrooms.size(); i++) {
                if (allClassrooms.get(i).equals(classroom)) {
                    classroomFound = classroom;
                    return classroomFound;
                }
            }
            // Eğer döngüde hiçbir eşleşme bulunmazsa, null kontrolü yapılır
            if (classroomFound == null) {
                throw new Exception("Classroom not found in the Classroom List");
            }
        } catch (Exception e) {
            System.out.println(e.getMessage());
            System.exit(1); // Programdan çıkış
        }
        return classroomFound; // Kod buraya hiç ulaşmaz ama syntax gereği yazıldı
    }

    // Check lecturer conflict for CourseSection.
    public boolean handleLecturerConflict(Lecturer lecturer, CourseSection courseSection) {
        boolean availability = true;
        int size = courseSections.size();
        for (int i = 0; i < size; i++) {
            if (courseSections.get(i).getLecturer() == lecturer) {
                for (int j = 0; i < courseSections.get(i).getTimeSlots().size(); j++) {
                    for (int k = 0; k < courseSection.getTimeSlots().size(); k++) {
                        if (courseSections.get(i).getTimeSlots().get(j).getTimeInterval() == courseSection
                                .getTimeSlots().get(k).getTimeInterval()) {
                            availability = false;
                        }
                    }
                }
            }
        }
        return availability;
    }
    // Returns a list of available days based on current course schedules and time intervals.
    public ArrayList<String> getAvailableDays(ArrayList<CourseSection> semesterCourses) {
        ArrayList<String> allDays = new ArrayList<>();
        allDays.add("Monday");
        allDays.add("Tuesday");
        allDays.add("Wednesday");
        allDays.add("Thursday");
        allDays.add("Friday");

        try {
            if (semesterCourses == null) {
                throw new IllegalArgumentException("Semester courses list cannot be null.");
            }

            // Tüm günler ve saat aralıklarını kontrol etmek için bir harita oluştur.
            HashMap<String, ArrayList<String>> dayToTimeIntervals = new HashMap<>();
            for (String day : allDays) {
                dayToTimeIntervals.put(day, new ArrayList<>(allTimeIntervals));
            }

            // Dolu olan zaman aralıklarını günlerden çıkar.
            for (CourseSection section : semesterCourses) {
                for (TimeSlot timeSlot : section.getTimeSlots()) {
                    String day = timeSlot.getDay();
                    String timeInterval = timeSlot.getTimeInterval();

                    // Eğer gün varsa ve zaman aralığı doluysa, onu kaldır.
                    if (dayToTimeIntervals.containsKey(day)) {
                        dayToTimeIntervals.get(day).remove(timeInterval);
                    }
                }
            }

            // Eğer bir günün tüm zaman aralıkları doluysa, o günü listeden çıkar.
            allDays.removeIf(day -> dayToTimeIntervals.get(day).isEmpty());
        } catch (Exception e) {
            System.err.println("Error while finding available days: " + e.getMessage());
        }

        return allDays;
    }

}

File: Iteration_3/DepartmentScheduler.py
from datetime import date
from CourseSection import CourseSection
from typing import List,Optional
from Staff import Staff

class DepartmentScheduler(Staff):
    def __init__(self, name: str = None, surname: str = None, birthdate: date = None, gender: str = None, ssn: str = None, course_sections=None, all_time_intervals=None):
        super().__init__(name, surname, birthdate, gender, ssn)
        self.__course_sections = course_sections or []
        self.__all_time_intervals = all_time_intervals or []
        self.__interface = None
        self.__all_classrooms = ["M2Z09", "M2Z10", "M2Z11", "M2Z12", "M2Z04"]
    def assign_time_slot_to_section(self, course_section, time_slot):
        if not course_section or not time_slot:
            raise ValueError("CourseSection or TimeSlot cannot be null.")
        time_slot.assign_time_slot(course_section)

    def assign_lecturer_to_section(self, course_section, lecturer):
        if not course_section or not lecturer:
            raise ValueError("CourseSection or Lecturer cannot be null.")
        if self.handle_lecturer_conflict(lecturer, course_section):
            course_section.set_lecturer(lecturer)

    def get_available_classrooms(self, time_slot):
        available_classrooms = self.__all_time_intervals.copy()
        for section in self.__course_sections:
            for slot in section.get_time_slots():
                if slot.get_time_interval() == time_slot.get_time_interval():
                    if time_slot.get_classroom() in available_classrooms:
                        available_classrooms.remove(time_slot.get_classroom())
        return available_classrooms

    def handle_time_conflict(self, semester_courses, day):
        try:
            available_times = self.__all_time_intervals.copy()
            for section in semester_courses:
                for time_slot in section.get_time_slots():
                    if time_slot.get_day() == day and time_slot.get_time_interval() in available_times:
                        available_times.remove(time_slot.get_time_interval())
            return available_times
        except Exception as e:
            print(str(e))
        
    def handle_classroom_conflict(self, day, time_interval):
        try:
            available_classrooms = self.__all_classrooms.copy()
            for section in self.get_course_sections():
                for time_slot in section.get_time_slots():
                    if time_slot.get_day() == day and time_slot.get_time_interval() == time_interval:
                        if time_slot.get_classroom() in available_classrooms:
                            available_classrooms.remove(time_slot.get_classroom())
            return available_classrooms
        except Exception as e:
            print(str(e))
            
    def handle_lecturer_conflict(self, lecturer, course_section):
        try:
            for section in self.__course_sections:
                if section.get_lecturer() == lecturer:
                    for slot in section.get_time_slots():
                        for new_slot in course_section.get_time_slots():
                            if slot.get_time_interval() == new_slot.get_time_interval():
                                return False
            return True
        except Exception as e:
            print(str(e))
            
    def get_available_days(self, semester_courses: list) -> list:
        all_days = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"]
        day_to_times = {day: self.__all_time_intervals.copy() for day in all_days}

        for section in semester_courses:
            for slot in section.get_time_slots():
                if slot.get_day() in day_to_times:
                    if slot.get_time_interval() in day_to_times[slot.get_day()]:
                        day_to_times[slot.get_day()].remove(slot.get_time_interval())

        return [day for day, times in day_to_times.items() if times]
    
    def semester_x_courses(self, x):
        semester_x_courses = []

        for course_section in self.__course_sections:
            if course_section.get_parent_course().get_semester() == x:
                semester_x_courses.append(course_section)
        return semester_x_courses

    def get_id(self):
        return super().get_ssn()

    def get_name(self):
        return super().get_name()

    def get_surname(self):
        return super().get_surname()

    def get_birthdate(self):
        return super().get_birthdate()

    def get_gender(self):
        return super().get_gender()
    
    def set_interface(self, interface):
        self.__interface = interface
        
    def initialize_interface(self):
        return self.__interface